Deserialize fewer SIL functions in -Onone

slavapestov opened the pull request:
Get mandatory inlining out of the linking business altogether. Mandatory inlining now only deserializes transparent functions. A separate mandatory linking pass deserializes shared definitions that we need to emit into the client.

slavapestov commented:
@swift-ci Please smoke test compiler performance

slavapestov committed:
SIL Linker: Only link non-public functions in LinkNormal mode

LinkNormal mode is used by the mandatory pipeline. It only needs
to deserialize what is necessary for code generation, that is
functions with shared linkage that must be emitted into the client.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@swift-ci Please smoke test compiler performance

slavapestov commented:
@swift-ci Please smoke test compiler performance

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@Swift-ci Please smoke benchmark

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
 //===----------------------------------------------------------------------===//
+//
+// The SIL linker walks the graph of all functions reachable from a starting
+// function, deserializing functions, vtables and witness tables.
```
"graph of all functions" => "call graph"?

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+// functions, to uncover optimization opportunities that might be missed
+// from deserializing late. The performance pipeline uses LinkAll mode.
+//
+// Note that in LinkAll mode, we deserialize all vtables and witness tables,
```
Can you make the Note bigger, i.e.:

```
*NOTE* In LinkAll mode
```

slavapestov replied:
Done.

gottesmm commented on lib/SILOptimizer/UtilityPasses/Link.cpp:
```diff
@@ -25,18 +25,27 @@ namespace {
 /// Copies code from the standard library into the user program to enable
 /// optimizations.
 class SILLinker : public SILModuleTransform {
+private:
```
private is not needed here.

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -58,7 +58,13 @@ bool SILLinkerVisitor::maybeAddFunctionToWorklist(SILFunction *F) {
+  // too.
+  if (isLinkAll() ||
+      hasSharedVisibility(F->getLinkage()) ||
+      F->getLinkage() == SILLinkage::HiddenExternal)
```
Two things:

1. Can't you do a check if this is public non ABI rather than doing this for all HiddenExternal functions?
2. Maybe split this up into separate predicates to make the flow clearer? I.e.:

```
// If we are supposed to link everything, just do it.
if (isLinkAll()) {
   ...
}

// Shared functions are emitted with link once odr linkage, so we want to import their bodies
// and re-emit them in the file. So if we have shared visibility, deserialize.
if (hasSharedVisibility(F->getLinkage())) {
   ...
}

// Finally, if we have a hidden external function with public non abi linkage, deserialize. This is because *snip* insert explanation here.
if (isPublicNonABI(F) && F->getLinkage() == SILLinkage::HiddenExternal) {
  ...
}

return false;
```

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -653,15 +653,6 @@ class MandatoryInlining : public SILModuleTransform {
                                SetFactory, SetFactory.getEmptySet(), CHA);
     }
 
-    // Make sure that we de-serialize all transparent functions,
```
Nice!

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -404,17 +404,22 @@ static SILFunction *getCalleeFunction(
+  if (CalleeFunction->empty()) {
+    // FIXME: Remove 'Mode'
+    if (Mode != SILOptions::LinkingMode::LinkNone)
+      AI.getModule().loadFunction(CalleeFunction);
```
Did you add the API loadFunction? I dislike that we have linkFunction and load function. I would rather have something more explicit about the transitive nature in the function names.

slavapestov replied:
I don't like the word "link" for what this code is doing at all, since its not really related to linking in the traditional sense. However, I'd like to punt this and the other renaming to my PR that removes LinkNone. Is that fair?

gottesmm commented on lib/SIL/Linker.h:
```diff
@@ -40,12 +40,16 @@ class SILLinkerVisitor : public SILInstructionVisitor<SILLinkerVisitor, bool> {
   LinkingMode Mode;
 
+  /// Whether any functions were deserialized.
+  bool LinkResult;
```
LinkResult => MadeChange?

gottesmm reviewed (commented):
Some initial comments.

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 //===----------------------------------------------------------------------===//
+//
```
Done.

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+//
+// - LinkNormal: The linker deserializes definitions that are required for
+//   code generation to proceed. This includes witness tables for imported
+//   conformances, and functions that have shared linkage.
```
I tried to rephase the paragraph a bit.

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+// The behavior of the linker is controlled by a LinkMode value. The LinkMode
+// has three possible values:
+//
+// - LinkNone: The linker does not deserialize anything. This is only used for
```
LinkNone is going away altogether in an upcoming PR, but I'm not touching it here.

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+// - LinkNone: The linker does not deserialize anything. This is only used for
+//   debugging and testing purposes, and never during normal operation.
+//
+// - LinkNormal: The linker deserializes definitions that are required for
```
I have another PR that removes LinkNone. How about in that PR I rename the enum (and move it somewhere more appropriate)

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+// SIL linker pass, which pulls in just enough to allow us to emit code, using
+// LinkNormal mode.
+//
+// In the performance pipeline, a performance SILLinker pass links all reachable
```
Done.

slavapestov commented on lib/SIL/Linker.cpp:
```diff
@@ -9,6 +9,39 @@
+//   code generation to proceed. This includes witness tables for imported
+//   conformances, and functions that have shared linkage.
+//
+// - LinkAll: All reachable functions are deserialized, including public
```
Fixed the second part.

slavapestov committed:
SIL Printer: Round-trip ImportDecls

We need this to update some other tests for SIL linker changes.
Now that we don't eagerly link everything in -Onone, we have to
pipe some test's SIL output through 'sil-opt -performance-linker',
but for that to work we must round-trip imports.

slavapestov commented:
@swift-ci Please smoke benchmark

lattner commented:
This is awesome Slava!  I'd really happy to see this.  Speeding up -O0 is important!

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -10,6 +10,46 @@
 //===----------------------------------------------------------------------===//
 
+
+/// \file The SIL linker walks the call graph beginning at a starting function,
```
Can you get rid of the two newlines before this line? Rest looks ok to me.

slavapestov committed:
SIL Optimizer: Update outdated comment concerning transparent functions

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@swift-ci Please test source compatibility

lattner commented:
awesome, what speedup do you see in the end?
